server/src/main/java/org/jboss/as/server/deployment/module/ModuleSpecProcessor.java

```diff
@@ -299,11 +304,12 @@ private void installAliases(final ModuleSpecification moduleSpecification, final
             dependencies.add(new ModuleDependency(moduleLoader, moduleIdentifier, false, false, false, false));
             ModuleDefinition moduleDefinition = new ModuleDefinition(alias, dependencies, spec);
 
-            final ValueService<ModuleDefinition> moduleSpecService = new ValueService<>(new ImmediateValue<>(moduleDefinition));
-            final ServiceBuilder sb = phaseContext.getServiceTarget().addService(moduleSpecServiceName, moduleSpecService);
+            final ServiceBuilder sb = phaseContext.getServiceTarget().addService(moduleSpecServiceName);
+            final Consumer<ModuleDefinition> moduleDefinitionConsumer = sb.provides(moduleSpecServiceName);
             sb.requires(deploymentUnit.getServiceName());
             sb.requires(phaseContext.getPhaseServiceName());
             sb.setInitialMode(Mode.ON_DEMAND);
+            sb.setInstance(new ModuleDefinitionService(moduleDefinitionConsumer, moduleDefinition));
```

Comment:
Why you are not using here the same approach used before? Cannot this be done via `org.jboss.msc.Service.newInstance(consumer, value)` ?

Reply:
It cannot be because ModuleDefinitionService is retrieved by other services via deprecated ServiceController.getValue() method and so it needs to implement legacy org.jboss.msc.service.Service interface.